File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        maxPathSumHelper(root);
        return ans;
    }
    public int maxPathSumHelper(TreeNode root){
        if(root == null)return Integer.MIN_VALUE;
        int l=maxPathSumHelper(root.left);
        int r=maxPathSumHelper(root.right);
        if(l<0)l=0;
        if(r<0)r=0;
        ans = Math.max(ans, l+r+root.val);
        return Math.max(l,r)+root.val;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_sum = float('-inf')

        def dfs(node):
            nonlocal max_sum
            if not node:
                return 0

            # compute max gain from left and right subtrees
            left = max(dfs(node.left), 0)   # ignore negative paths
            right = max(dfs(node.right), 0)

            # path that passes through current node
            current_max = node.val + left + right
            max_sum = max(max_sum, current_max)

            # return max gain if path continues through parent
            return node.val + max(left, right)

        dfs(root)
        return max_sum
